test/rules-adobe-fullstory.spec.ts

```diff
@@ -1,56 +1,94 @@
 import 'mocha';
 
-import { rules } from '../examples/rules/adobe-fullstory.json';
-import { basicAppMeasurement } from './mocks/adobe';
-import {
-  expectEqual, expectMatch, expectRule, expectFS, setupGlobals, ExpectObserver, expectUndefined, expectGlobal,
-  expectNoCalls,
-} from './utils/mocha';
-
-describe('Adobe to FullStory rules', () => {
-  beforeEach(() => setupGlobals([
-    ['s', basicAppMeasurement],
-    ['_dlo_rules', rules],
-  ]));
-
-  afterEach(() => {
-    ExpectObserver.getInstance().cleanup();
-  });
+import { expectEqual, expectMatch, expectUndefined } from './utils/mocha';
+import { RulesetTestHarness, getRulesetTestEnvironments } from './utils/ruleset-test-harness';
+import { basicAppMeasurement, AppMeasurement } from './mocks/adobe';
 
-  it('it should send just Adobe eVars to FS.event', () => {
-    ExpectObserver.getInstance().create(
-      { rules: [expectRule('fs-event-adobe-evars')], readOnLoad: true },
-    );
+import '../rulesets/adobe-app-measurement.js';
 
-    const [eventName, payload] = expectFS('event');
-    expectEqual(eventName, 'Adobe eVars');
-    expectMatch(basicAppMeasurement, payload, 'eVar1', 'eVar10', 'eVar20', 'eVar50', 'eVar60');
-    expectUndefined(payload, 'prop1', 'pageName');
-  });
+const adobeRulesKey = '_dlo_rules_adobe_am';
+const adobeRules = (window as Record<string, any>)[adobeRulesKey];
 
-  it('should not send empty eVar objects to FS.event', () => {
-    const s = expectGlobal('s');
+declare global {
+  // eslint-disable-next-line no-var, vars-on-top
+  var s: AppMeasurement;
+}
 
-    // Set all eVars to undefined to simulate an empty object
-    Object.keys(s)
-      .filter((key) => key.startsWith('eVar'))
-      .forEach((key) => { s[key] = undefined; });
+describe('Ruleset: Adobe to FullStory rules', () => {
+  getRulesetTestEnvironments().forEach((testEnv) => {
+    describe(`test environment: ${testEnv.name}`, () => {
+      let testHarness: RulesetTestHarness;
 
-    ExpectObserver.getInstance().create(
-      { rules: [expectRule('fs-event-adobe-evars')], readOnLoad: true },
-    );
+      afterEach(async () => {
+        await testHarness.tearDown();
+      });
 
-    expectNoCalls(expectGlobal('FS'), 'event');
-  });
+      after(async () => {
+        await testEnv.tearDown();
+      });
+
+      it('sends Adobe eVars and props to FS.event', async () => {
+        testHarness = await testEnv.createTestHarness(adobeRules, {
+          s: {
+            eVar1: '',
```

Comment:
I dont think the verbosity is a big deal here, why do we need so many parameters like 1, 10, 20, 50, 60? do these values have special meaning? (sorry still learning more about DLO)

Reply:
Good questions!

> why do we need so many parameters like 1, 10, 20, 50, 60?

We don't need this params specifically. This is meant to test a range of `eVar` values since Adobe App Measurement data layers go from `eVar1` to `eVar250` and `prop` values go from `prop1` to `prop75`. These parameters give us some confidence that we're tracking all `eVar` and `prop` values without exhaustively setting initial state for all 250 `eVar`s and all 75 `prop`s, changing values, and verifying all are present in the event payload. Definitely open to other suggestions here.

> why doesn't the other test need to verify all these properties are also undefined?

I saw the first test as verifying the majority of `eVar`s and `prop`s are being tracked and the second test as a quick confidence check that setting to `undefined` is ignored (this wasn't always the behavior). We could be just as exhaustive in this test if the consistency is more expected or is otherwise valuable.